hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java

```diff
@@ -0,0 +1,451 @@
+        if (target == null) {
+            throw new MutationDisallowedException(
+                    "Cannot invoke operations on a CRDT because the cluster does not contain any data members");
+        }
+        final ClientMessage response = invokeAddInternal(-1, true, EMPTY_ADDRESS_LIST, null, target);
+        final PNCounterAddCodec.ResponseParameters resultParameters = PNCounterAddCodec.decodeResponse(response);
+        updateObservedReplicaTimestamps(resultParameters.replicaTimestamps);
+        return resultParameters.value;
+    }
+
+    @Override
+    public long getAndIncrement() {
+        final Address target = getCRDTOperationTarget(EMPTY_ADDRESS_LIST);
+        if (target == null) {
+            throw new MutationDisallowedException(
+                    "Cannot invoke operations on a CRDT because the cluster does not contain any data members");
+        }
+        final ClientMessage response = invokeAddInternal(1, true, EMPTY_ADDRESS_LIST, null, target);
+        final PNCounterAddCodec.ResponseParameters resultParameters = PNCounterAddCodec.decodeResponse(response);
+        updateObservedReplicaTimestamps(resultParameters.replicaTimestamps);
+        return resultParameters.value;
+    }
+
+
+    /**
+     * Transforms the list of replica logical timestamps to a vector clock instance.
+     *
+     * @param replicaLogicalTimestamps the logical timestamps
+     * @return a vector clock instance
+     */
+    private VectorClock toVectorClock(List<Entry<String, Long>> replicaLogicalTimestamps) {
+        final VectorClock timestamps = new VectorClock();
+        for (Entry<String, Long> replicaTimestamp : replicaLogicalTimestamps) {
+            timestamps.setReplicaTimestamp(replicaTimestamp.getKey(), replicaTimestamp.getValue());
+        }
+        return timestamps;
+    }
+
+    /**
+     * Adds the {@code delta} and returns the value of the counter before the
+     * update if {@code getBeforeUpdate} is {@code true} or the value after
+     * the update if it is {@code false}.
+     * It will invoke client messages recursively on viable replica addresses
+     * until successful or the list of viable replicas is exhausted.
+     * Replicas with addresses contained in the {@code excludedAddresses} are
+     * skipped. If there are no viable replicas, this method will throw the
+     * {@code lastException} if not {@code null} or an {@link MutationDisallowedException}
+     * if the {@code lastException} is {@code null}.
+     *
+     * @param delta             the delta to add to the counter value, can be negative
+     * @param getBeforeUpdate   {@code true} if the operation should return the
+     *                          counter value before the addition, {@code false}
+     *                          if it should return the value after the addition
+     * @param excludedAddresses the addresses to exclude when choosing a replica
+     *                          address, may be {@code null}
```

Comment:
Seems like it can't be `null` actually, this will produce NPE.

Reply:
Leftover, thanks

Reply:
Fixed.